### PR TITLE
changed the way we calculate appx. solution size.

### PR DESCRIPTION
**Pull request by heejaechang**
previously we delayed calculation as much as we can. basically after everything else. that made our lazy persistent service to kick in (enabled) almost at the end.

this should let it be enabled as soon as possible.

### CODE REVIEW

**Comment by heejaechang** [comment]
@pharring @rchande can you take a look? thank you

File: src/VisualStudio/CSharp/Test/CSharpVisualStudioTest.csproj [hunk 1]
```diff
@@ -116,23 +116,55 @@
   <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'Release|AnyCPU'">
   </PropertyGroup>
   <ItemGroup>
-    <Reference Include="envdte, Version=8.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a"><Private>false</Private></Reference>
-    <Reference Include="envdte80, Version=8.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a"><Private>false</Private></Reference>
+    <Reference Include="envdte, Version=8.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a">
+      <Private>false</Private>
```
Comment (rchande) [inline src/VisualStudio/CSharp/Test/CSharpVisualStudioTest.csproj]: Unrelated?

**Comment by rchande** [comment]
@heejaechang What happens if GetSolutionSize is called before every document has been analyzed?

File: src/VisualStudio/Core/Def/Implementation/SolutionSize/SolutionSizeTracker.cs [hunk 2]
```diff
@@ -57,79 +56,29 @@ public Task NewSolutionSnapshotAsync(Solution solution, CancellationToken cancel
...
-                }
-
-                var sum = 0L;
-                foreach (var document in project.Documents)
+                // remove old size
+                long size;
+                if (_map.TryGetValue(document.Id, out size))
                 {
-                    sum += await GetDocumentSizeAsync(document, cancellationToken).ConfigureAwait(false);
+                    _size -= size;
```
Comment (pharring) [inline src/VisualStudio/Core/Def/Implementation/SolutionSize/SolutionSizeTracker.cs]: If ```size``` is the same as ```tree.Length```, then we should return early.

Also, should ```_size``` be updated via ```Interlocked.Add```? i.e. Does this code run in parallel?

Comment (heejaechang) [inline src/VisualStudio/Core/Def/Implementation/SolutionSize/SolutionSizeTracker.cs]: this code never runs parallel.

File: src/VisualStudio/CSharp/Test/CSharpVisualStudioTest.csproj (hunk 1 above)
Comment (heejaechang) [inline src/VisualStudio/CSharp/Test/CSharpVisualStudioTest.csproj]: unrelated. somehow reformated.

**Comment by heejaechang** [comment]
@rchande this is not something give you final solution size, it is a service that answers solution size (of solution crawler's point of view) at the point of query.

File: src/VisualStudio/Core/Def/Implementation/SolutionSize/SolutionSizeTracker.cs (hunk 2 above)
Comment (pharring) [inline src/VisualStudio/Core/Def/Implementation/SolutionSize/SolutionSizeTracker.cs]: If not (guaranteed to run sequentially), then do we actually need ```ConcurrentDictionary``` or would a regular ```Dictionary<K,V>``` suffice?

**Comment by heejaechang** [comment]
@pharring @rchande can you take a look?

File: src/VisualStudio/Core/Def/Implementation/SolutionSize/SolutionSizeTracker.cs [hunk 3]
```diff
@@ -49,87 +55,44 @@ public Task NewSolutionSnapshotAsync(Solution solution, CancellationToken cancel
...
-                _size = _map.Values.Sum();
-            }
+                // getting tree is cheap since tree always stays in memory
+                var tree = await document.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+                var length = tree.Length;
 
-            private static async Task<long> GetProjectSizeAsync(Project project, CancellationToken cancellationToken)
-            {
-                if (project == null)
+                _map.AddOrUpdate(document.Id, id =>
```
Comment (pharring) [inline src/VisualStudio/Core/Def/Implementation/SolutionSize/SolutionSizeTracker.cs]: I don't think you can use AddOrUpdate safely because the delegates (addValueFactory and updateValueFactory) may be called more than once (if an update/add fails and has to be retried) and these two delegates have side effects (updating _size).

Unfortunately, you can't use the return value from AddOrUpdate to update _size because you need to know the old size too.

So, I think you'll have to write this out in full using TryGet / TryUpdate / TryAdd.

Comment (heejaechang) [inline src/VisualStudio/Core/Def/Implementation/SolutionSize/SolutionSizeTracker.cs]: ah, looked at the source code. you are right. let me change it again.

**Comment by heejaechang** [comment]
@pharring updated

**Comment by pharring** [comment]
:+1: thanks.
